src/dependencies/useStoreDependency.ts

```diff
@@ -1,72 +1,105 @@
-  const ref = React.useRef(value);
+export function useCurrent<ValueType>(value: ValueType): RefObject<ValueType> {
+  const ref = useRef(value);
   ref.current = value;
   return ref;
 }
 
-function useStoreDependency<Props>(
-  dependency: Dependency,
-  props: Props,
-  dispatcher: Dispatcher<any> = getDispatcherInstance()
+export function _useDispatchSubscription<
+  Props,
+  DependencyMapType extends DependencyMap,
+  DependenciesType = {
+    [key in keyof DependencyMapType]: GetDependencyType<DependencyMapType[key]>
+  }
+>(
+  dependencyMap: DependencyMapType,
+  currentProps: RefObject<Props>,
+  dispatcher: Dispatcher<any>,
+  dependencyValue: DependenciesType,
+  setDependencyValue: Dispatch<SetStateAction<DependenciesType>>
 ) {
-  enforceDispatcher(dispatcher);
-
-  const [dependencyValue, setDependencyValue] = React.useState(
-    calculate(dependency, props)
-  );
-
-  const currProps = useCurrent(props);
-
-  React.useEffect(() => {
-    const dependencyMap = { dependency };
+  useEffect(() => {
     const dependencyIndex = makeDependencyIndex(dependencyMap);
     const dispatchToken: string = dispatcher.register(
       handleDispatch.bind(
         null,
         dispatcher,
         dependencyIndex,
         (entry: DependencyIndexEntry) => {
-          const {
-            dependency: newValue,
-          }: SingleDependency = calculateForDispatch(
-            dependencyMap,
-            entry,
-            currProps.current
-          );
+          const newValue = calculateForDispatch<
+            Props,
+            Partial<typeof dependencyMap>,
+            DependencyMapType
+          >(dependencyMap, entry, currentProps.current);
           if (!shallowEqual(newValue, dependencyValue)) {
-            setDependencyValue(newValue);
+            setDependencyValue((newValue as unknown) as DependenciesType);
           }
         }
       )
     );
     return () => {
       dispatcher.unregister(dispatchToken);
     };
-  }, [dispatcher, dependencyValue, dependency, currProps]);
+  }, [
+    currentProps,
+    dependencyMap,
+    dependencyValue,
+    dispatcher,
+    setDependencyValue,
+  ]);
+}
+
+function useStoreDependency<Props, DepType>(
+  dependency: Dependency<DepType>,
+  props?: Props,
+  dispatcher: Dispatcher<any> = getDispatcherInstance()
+): DepType {
+  enforceDispatcher(dispatcher);
+
+  const [dependencyValue, setDependencyValue] = useState({
+    dependency: calculate(dependency, props),
+  });
+
+  const currProps = useCurrent(props);
+
+  const dependencyMap = useMemo(() => ({ dependency }), [dependency]);
+
+  _useDispatchSubscription(
+    dependencyMap,
+    currProps,
+    dispatcher,
+    dependencyValue,
+    setDependencyValue
+  );
 
   const newValue = calculate(dependency, props);
-  if (!shallowEqual(newValue, dependencyValue)) {
-    setDependencyValue(newValue);
+  if (!deepEqual(newValue, dependencyValue.dependency)) {
```

Comment:
Curious why the change here from shallow to deep?

Reply:
whoops, copypasta from the new version of `connect`. `connect` requires the `deepEqual` because the state in that component is a DependencyMap, so we need to check that the entire map is the same and trigger the setState if any of the sub-objects is meaningfully different, but in `useStoreDependency` it's a single result so it can be a shallowEqual. I'll update